### PR TITLE
llama_get_embeddings_mean_pooled

### PR DESCRIPTION
**Pull request by mgrosso**
## the problem being solved here

described in https://github.com/ggerganov/llama.cpp/issues/6754 , which was created for the PR.

### It's hard to use llama.cpp on the state of the art embedding models when using language bindings

Several leading models on the MTEB embeddings leaderboard use "large" transformer models that have been fine tuned to produce embeddings instead of or in addition to generating tokens, and in all 3 cases an instruction sequence of tokens is prepended to the query. Examples include GritLM, e5-mistral, and echo-mistral. (links below). While echo-mistral was evaluated using just the last token of the last hidden layer, both GritLM and e5-mistral have been trained and evaluated using mean pooling on the embeddings of the last hidden layer. In the GritLM example we implement that in custom c++ outside of the device.

When attempting to utilize the same functionality with a language binding like `llama_cpp.rb` one quickly finds that you have a choice between:

1. doing the mean polling in the binding language, which is not fast in the case of ruby unless you implement a whole other fast math layer in addition to gguf, or
2. adding llama.cpp mean polling into the binding layer, where it really doesn't seem to belong, or
3. add external mean pooling to llama.cpp at the risk of duplicating the BERT specific mean pooling that is currently available.
4. extending the mean pooling available to non-BERT models, but in a way that allows the application control over how many instruction tokens to skip.

Note that the token skipping requirements would be more complicated if the echo-mistral approach is combined with mean pooling.

I chose option 3 here, because options 1 & 2 seem wrong and I don't know the project well enough to attempt 4, never mind make option 4 flexible enough.

## the change

I refactored the off device mean pooling from examples/gritlm/gritlm.cpp to `llama_get_embeddings_mean_pooled` where it can be picked up by language bindings, or potentially used in a server endpoint.

## testing

I ran `./gritlm --model ./models/ggml-gritlm-7b-q4_k.gguf` before and after the change. I got the same similarities between the vectors before and after:

        Cosine similarity between "Bitcoin: A Peer-to-Peer Electronic Cash System" and "A purely peer-to-peer version of electronic cash w" is: 0.601
        Cosine similarity between "Bitcoin: A Peer-to-Peer Electronic Cash System" and "All text-based language problems can be reduced to" is: 0.098
        Cosine similarity between "Generative Representational Instruction Tuning" and "A purely peer-to-peer version of electronic cash w" is: 0.124
        Cosine similarity between "Generative Representational Instruction Tuning" and "All text-based language problems can be reduced to" is: 0.552

### CODE REVIEW

**Comment by ngxson** [comment]
In the `struct llama_context_params`, there's a param called `llama_pooling_type` which can be set to `LLAMA_POOLING_TYPE_MEAN`. Have you tried it out?

**Comment by iamlemec** [comment]
This seems like a useful idea! As you describe, it seems like a tricky thing to satisfy all the various constraints. On the `llama-cpp-python` side there have also been reports of issues and confusion with getting embeddings from generative or hybrid models.

I agree that number 3 seems closest to optimal. It might even be desirable to implement the pooling logic as here and have all models, including the BERTs, use that as well. The drawback would be that the pooling computations would be happening on device. I'm not sure how much of a performance hit this would be, but doing it on CPU will be in principle slower, and it would increase the amount of data copied from device to CPU by a factor of the average sequence length. But in this end, this may well be negligible compared to everything else. To get around this, we could tack on the pooling ops to the computational graphs, but I'm not sure if this is "allowed" currently.

Only other thing is it looks like this PR currently pools all of the tokens in a batch together. In general, it would be desirable to pool tokens together by sequence within a batch. This is what the BERTs currently do, and it allows you to efficiently pack multiple sequences into each batch.

@ngxson `LLAMA_POOLING_TYPE_MEAN` is only used by BERT models in `build_bert`. All other model types would ignore it.

**Comment by mgrosso** [comment]
> In the `struct llama_context_params`, there's a param called `llama_pooling_type` which can be set to `LLAMA_POOLING_TYPE_MEAN`. Have you tried it out?

I have tried that. it fails near the end of decode on `ASSERT(lctx->inp_mean)` if I remember it right, since that tensor is only built for BERT models,  as @iamlemec points out above. 

Note that this new API would be forward compatible with any future effort to bring `inp_mean` into other model's builds, assuming `skip_tokens` was `0` and `batch.logits[...]` was set false for instruction tokens, as documented in the API comments.

I personally couldn't see how to use the state of the art embedding models in `llama_cpp.rb` without computing `n_embd * (batch.n_tokens + 1)` floating point operations with interpreted ruby, or else bringing another numerical computing system into the project besides ggml, neither of which was desirable. I'm right now figuring out how to manage this and future patches to both projects into my build system, so I'm motivated to help find a solution the core team thinks is worth merging.

In addition to the `llama-cpp-python` issues just mentioned, it looks to me like a c# binding project is struggling with the same issue, based on reading their issue https://github.com/SciSharp/LLamaSharp/issues/673 (thanks @martindevans for flagging that)

**Comment by ngxson** [comment]
Thanks for the info. You’re right @iamlemec , the current LLAMA_POOLING_TYPE_MEAN only works with BERT. However, as you pointed out, the downside of llama_get_embeddings_mean_pooled introduced in this PR is that the calculation is offloaded to CPU (plus some overhead of copying data between GPU - CPU). I think it’s better to ask for advice from @slaren about this subject.

Another thing worth mentioning is that this new API may be confused for users, since we’re now having 2 ways to get the mean pooling. Personally, I would prefer to patch the current LLAMA_POOLING_TYPE_MEAN instead. Not necessary implement the GPU part, but maybe just patch to firstly "make it work" then we will see if we can improve it.

**Comment by mgrosso** [comment]
> Thanks for the info. You’re right @iamlemec , the current LLAMA_POOLING_TYPE_MEAN only works with BERT. However, as you pointed out, the downside of llama_get_embeddings_mean_pooled introduced in this PR is that the calculation is offloaded to CPU (plus some overhead of copying data between GPU - CPU). I think it’s better to ask for advice from @slaren about this subject.
> 
> Another thing worth mentioning is that this new API may be confused for users, since we’re now having 2 ways to get the mean pooling. Personally, I would prefer to patch the current LLAMA_POOLING_TYPE_MEAN instead. Not necessary implement the GPU part, but maybe just patch to firstly "make it work" then we will see if we can improve it.

@ngxson thanks for the thoughtful review. I'll close this for now and see if I can come up with a fix that doesn't introduce a new API. If not, hopefully the supporting issue helps document the problem.

One question... suppose I tucked my fix into a `if` block on `llama_get_embedding` and of course another conditional to avoid the assert on `inp_mean`. Where would be a reasonable place to write out the `n_embd` floats? I suppose I could add some better named version of `inp_mean_for_non_bert` to `llama_context`?  because if the memory doesn't go in the context, and the `llama_batch` seems to be for inputs only, I'm not sure where else it might go.  

In terms of a more integrated fix I'd appreciate guidance from @slaren or anyone else on how the context level choice of whether to do embeddings or not (GritLM) should drive the allocation of a ggml tensor in the model level `build_llama`. It's only 16k for the biggest embeddings right now so I assume it could just be unconditional, but I'm not far enough along in my GGML and llama.cpp journey to know the implications of adding a conditionally used output unconditionally... or even conditionally for that matter. I don't want to introduce a performance regression in inferencing or some other rake in the grass.

**Comment by ngxson** [comment]
@mgrosso Sorry for the late response. Yes, I think we can start by patching `llama_get_embedding` to internally work with `inp_mean_for_non_bert` if `LLAMA_POOLING_TYPE_MEAN` is set. The implementation can be more or less like the current proposal (calculating mean value using CPU). In the future, we will see if this can be offloaded to GPU.

**Comment by sirus20x6** [comment]
I would love support for https://huggingface.co/Salesforce/SFR-Embedding-Mistral which is based on e5
